File: src/frontend/src/app/upload-batch.ts
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map, Observable } from 'rxjs';

export enum UploadBatchStatus {
  PENDING = 'PENDING',
  PROCESSING = 'PROCESSING',
  COMPLETED = 'COMPLETED',
  PARTIAL = 'PARTIAL',
  FAILED = 'FAILED',
}

export const UploadBatchStatusLabel: Record<UploadBatchStatus, string> = {
  [UploadBatchStatus.PENDING]: 'Pendente',
  [UploadBatchStatus.PROCESSING]: 'Processando',
  [UploadBatchStatus.COMPLETED]: 'Completado',
  [UploadBatchStatus.PARTIAL]: 'Enviado Parcialmente',
  [UploadBatchStatus.FAILED]: 'Com Erro',
};

export interface GetUploadBatchesResponse {
  id: string; // Guid
  name: string;
  uploadBatchStatus: UploadBatchStatus;
  createdAt: Date;
  completedAt: Date | null;
  totalFile: number;
}

export interface CreateUploadBatchPayload {
  name: string;
  filesInfo: Array<{
    objectKey: string;
    originalFileName: string;
    mimeType: string;
    size: number;
    title: string;
    tags: string[];
    metadata: Record<string, unknown> | null;
  }>;
}

export interface CreateUploadBatchResponse {
  id: string;
}

export interface GetUploadBatchStatusResponse {
  id: string;
  status: UploadBatchStatus;
  timedOut: boolean;
}

@Injectable({
  providedIn: 'root',
})
export class UploadBatchService {
  private readonly batchesApiUrl = 'api/upload-batches';
  private readonly createBatchApiUrl = 'api/upload-batch';

  constructor(private httpClient: HttpClient) {}

  getAll(): Observable<GetUploadBatchesResponse[]> {
    return this.httpClient.get<GetUploadBatchesResponse[]>(this.batchesApiUrl);
  }

  /**
   * Cria um lote de upload com arquivos já uploadados no storage
   * @param payload Contém os dados de criação do lote e referências dos arquivos já enviados
   */
  create(payload: CreateUploadBatchPayload): Observable<CreateUploadBatchResponse> {
    console.log(`[UploadBatchService] Criando lote: ${payload.name}`);

    return this.httpClient
      .post(this.createBatchApiUrl, payload, {
        observe: 'response',
      })
      .pipe(
        map((response) => ({ id: this.extractBatchId(response) }))
      );
  }

  getStatus(id: string): Observable<GetUploadBatchStatusResponse> {
    return this.httpClient.get<GetUploadBatchStatusResponse>(
      `${this.batchesApiUrl}/${id}/status`
    );
  }

  private extractBatchId(response: HttpResponse<unknown>): string {
    const bodyId = this.extractIdFromBody(response.body);
    if (bodyId) {
      return bodyId;
    }

    const location = response.headers.get('location');
    if (location) {
      const idFromLocation = location.split('/').pop();
      if (idFromLocation) {
        return idFromLocation;
      }
    }

    throw new Error('Nao foi possivel identificar o id do lote criado.');
  }

  private extractIdFromBody(body: unknown): string | null {
    if (typeof body === 'string' && body.trim().length > 0) {
      return body;
    }

    if (body && typeof body === 'object' && 'id' in body) {
      const id = (body as { id?: unknown }).id;
      if (typeof id === 'string' && id.trim().length > 0) {
        return id;
      }
    }

    return null;
  }
}

File: src/FileFlow.Data/Entities/UploadBatch.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json;

namespace FileFlow.Data.Entities;

[Table("upload_batches")]
public class UploadBatch : Entity
{
    [Column("name")]
    [Required]
    [MaxLength(255)]
    public string Name { get; private set; } = null!;

    [Column("status", TypeName = "varchar(50)")]
    [Required]
    public UploadBatchStatus Status { get; private set; }

    [Column("created_at")]
    [Required]
    public DateTime CreatedAt { get; private set; }

    [Column("completed_at")]
    public DateTime? CompletedAt { get; private set; }

    [InverseProperty(nameof(MediaAsset.UploadBatch))]
    private readonly List<MediaAsset> _mediaAssets = [];

    public IReadOnlyCollection<MediaAsset> MediaAssets => _mediaAssets;

    public static UploadBatch Create(string name)
    {
        return new UploadBatch
        {
            Id = Guid.NewGuid(),
            Name = name,
            Status = UploadBatchStatus.PENDING,
            CreatedAt = DateTime.UtcNow,
            CompletedAt = null,
        };
    }

    public MediaAsset AddMediaAsset(
        string originalFileName,
        string mimeType,
        long size,
        string title,
        List<string>? tags,
        JsonDocument? metadata)
    {
        var mediaAsset = MediaAsset.Create(
            Id,
            originalFileName,
            mimeType,
            size,
            title,
            tags,
            metadata);

        _mediaAssets.Add(mediaAsset);

        return mediaAsset;
    }

    public void MarkCompleted(DateTime completedAt, UploadBatchStatus status = UploadBatchStatus.COMPLETED)
    {
        CompletedAt = completedAt;
        Status = status;
    }

    protected UploadBatch() { }
}

public enum UploadBatchStatus
{
    PENDING,
    PROCESSING,
    COMPLETED,
    PARTIAL,
    FAILED,
}